src/checkers/Cashflow.ml

```diff
@@ -1561,6 +1561,27 @@ module ScillaCashflowChecker
             new_ctr_tag_map,
             new_changes || acc_changes))
         
+  let cf_tag_field field param_env field_env ctr_tag_map =
+    let (x,ftype,e) = field in
+    (* Bind should have same functionality except that the environment should be field_env and not local_env *)
+    (* problem: the tokens1 in field env is never known to the cf_tag_expr and so it cannot add it's new tag  *)
+    let local_env = field_env in
+
+    let x_tag = lookup_var_tag x local_env in (* expected tag is the tag in the current env  *)
+    let e_local_env = AssocDictionary.remove (get_id x) local_env in
+    let ((_, (new_e_tag, _)) as new_e, new_param_env, new_local_env, new_ctr_tag_map, e_changes) =
+      cf_tag_expr e x_tag param_env e_local_env ctr_tag_map in
+    let new_x_tag = lub_tags x_tag new_e_tag in (* new field tag based on LUB *)
+    let new_x = update_id_tag x new_x_tag in (* new field = tokens1 *)
+
+    let new_field_env = AssocDictionary.insert (get_id new_x) (get_id_tag new_x) new_local_env in
```

Comment:
Once you use the correct local environment, you won't need to remove `x` from the field environment, and then you can use `AssocDictionary.update` instead. Even better, use the helper function `update_var_tag`.